### ISSUE TITLE
configure .cache via $XDG_CONFIG_DIR

### ISSUE THREAD

**Issue by bukzor** [pytest-dev/pytest#1089]
Forked from #1029 

---

a) A very convenient, and standard, way to configure tools' cache location is to set $XDG_CACHE_DIR. Many of my project's fixtures do this for other reasons already. Similarly, the most convenient, and standard, way to set "basetmp" would be $TMPDIR.

b) "something a CI system can easily pick up" is exactly these environment variables. Many systems will already support this, and those that don't will support injecting environment variables.

c) "non-containered ci" will either support ~/.cache correctly, or set $XDG_CACHE_DIR, because this is necessary for other tools that use the standard. Again, this is already handled in many of my projects because of other tools.

**Comment by RonnyPfannschmidt** [comment]
is there a good package that helps finding those paths and helps clearing unused caches

after all if the cache folder is outside of the working directory,
its lifetime is unrelated to the working directory

imagine for example a unaware ci system that makes one folder per build could easily DOS the ci server due to filling up ~/.cache of the ci user

with the current mode of operation, cache and working directory are simply co-located and share a lifetime, so data management is easy, the cache is gone when one deletes the working directory

im not per se opposed to following XDG, but it makes lifetimes of file-system objects more tricky on our case and seems not really available on windows either
(its my personal impression that xdg works only well on posix boxes with a typical linux/bsd setup)

**Comment by bukzor** [comment]
Those are good points...

I don't have a good answer. The applications I've seen don't worry about cleaning up like that.
It would work for me if we use the current behavior when XDG_CACHE_DIR doesn't exist.

@asottile

**Comment by bukzor** [comment]
This one at least helps find a good platform-specific cache dir: https://pypi.python.org/pypi/appdirs/1.4.0

**Comment by bukzor** [comment]
This one helps manage expiry for a file-based cache.
Making one of the keys be the working directory would make it work.

http://pythonhosted.org/pyfscache/

**Comment by RonnyPfannschmidt** [comment]
i shorty investigated appdirs and pyfscache, unfortunately both are unsuitable

appdirs seems bug-ridden
pyfscache is for something entirely different not fitting the use-case

**Comment by RonnyPfannschmidt** [comment]
how about introducing a ini value/env variable to select the cache backend,
which defaults to worktree, but one can choose xdg/disabled?

@hpk42 oppinions?

**Comment by asottile** [comment]
Personally I'd rather default to xdg instead of needing to add .cache to the gitignore in every one of my projects

**Comment by RonnyPfannschmidt** [comment]
while i agree that it is a nice default, we cant do a change like that befor 3.0

also before we can do it propperly we need to come up with a solid way to clean things up

**Comment by asottile** [comment]
Maybe introducing the cache plugin defaulted to on should have been in 3.0 too :/

**Comment by RonnyPfannschmidt** [comment]
since its a normal feature addition, it doesnt warrant a release that large

**Comment by s-trooper** [comment]
i hate ".cache" folder for aesthetic reason. my projects folders are already populated by ide, git, python and others "support folders". sometimes i have more folders than actually python program files! please make ".cache" location configurable!

**Comment by nicoddemus** [comment]
> how about introducing a ini value/env variable to select the cache backend

Anyone is against this? While it may not be the best option, it is simple to implement and backward compatible.

It should support variable expansion and default to `.cache` for backward compatibility. This would allow users to change it like this:

```
[pytest]
cache_dir=$XDG_CACHE_DIR
```

Or using an environment variable:

```
PYTEST_CACHE_DIR=/some/path
```

**Comment by RonnyPfannschmidt** [comment]
I'm against the most easy variant, I want a simple solution

**Comment by bukzor** [comment]
@RonnyPfannschmidt I've found that the function you'd need exists in pip.utils:

https://github.com/pypa/pip/blob/develop/pip/utils/appdirs.py#L13

**Comment by nicoddemus** [comment]
How about we add a configuration option that decides between the two systems? `cache_location`, which can be `local` (`<rootdir>/.cache`, the default) or `system` (using the heuristics suggested by @bukzor)? This could be added in `2.9`.

If we make a poll and people wants to change `system` as the default, we can announce that the default will change in `2.10`, giving plenty of time for people to change to their desired setting while in `2.9` and not get caught off guard.

BTW, I'm just throwing some ideas so we can reach some consensus, I don't mind having the `.cache` directory in pytest's rootdir: I configured my git-global-ignore file to always ignore it.

**Comment by RonnyPfannschmidt** [comment]
:+1:

**Comment by bukzor** [comment]
For the implementation of `system`, I'd probably just import and use
pip.utils.
Would you all be opposed to a dependency on pip from pytest?
I imagine ~100% of python installations have pip installed already.

**Comment by bukzor** [comment]
Actually, the pip function seems to be closely related to the appdirs package implementation.
I'm not sure which is derived from which though.

https://github.com/ActiveState/appdirs/blob/master/appdirs.py#L257

**Comment by nicoddemus** [comment]
IMHO I would just copy the function over, licensing permitting... depend on the entire `pip` library just for the sake of a small function like that is not worth it, I think, as I'm not sure how stable is that API.

**Comment by RonnyPfannschmidt** [comment]
I'd go for publishing a minimalistic lib that implements it

**Comment by antoche** [comment]
Hi,
Is the recommended solution to add .cache to every project's .gitignore? Or is there a consensus on a solution for this issue?
Thanks,
A.

**Comment by RonnyPfannschmidt** [comment]
@antoche for now yes

**Comment by DuncanBetts** [comment]
You can also add .cache to your global gitignore, rather than on a per project basis — as niccoddemus comments [above](https://github.com/pytest-dev/pytest/issues/1089#issuecomment-166977102).

I believe the command is:

`$ git config --global core.excludesfile .cache`

**Comment by nicoddemus** [comment]
There's a `cache_dir` ini option which is now available in `3.2`. 

Should we change the default of that option to `$XDG_CONFIG_DIR` (if defined) in future pytest versions?

**Comment by RonnyPfannschmidt** [comment]
the config dir is under all circumstances always the wrong folder for the cache

we should use the cache home as according to the spec,

and we should take a look at using the appdirs lib to facilitate those details

**Comment by bukzor** [comment]
@nicoddemus: No, the default should be [$XDG_CACHE_DIR](https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html), which in turn has a default of $HOME/.cache.

> $XDG_CACHE_HOME defines the base directory relative to which user specific non-essential data
> files should be stored. If $XDG_CACHE_HOME is either not set or empty, a default equal to
> $HOME/.cache should be used.

If you want to support macos and windows in a natural way, you'll want a library for this. I think this is the right solution: https://github.com/ActiveState/appdirs#the-problem

**Comment by RonnyPfannschmidt** [comment]
@bukzor thanks for the clarification and correcting my mistake

**Comment by nicoddemus** [comment]
Thanks @bukzor.

Hmm just realized that with this change the cache will no longer be per-repository, but will be global per-user. This might be a problem because plugins assume a per-repository cache, for example pytest's `--lf` option.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus borgbackup for example uses the repository path as extra path component into the per user caches in order to keep things appart

**Comment by bukzor** [comment]
I've used that strategy myself with success, but I only care about Unixen.
You'll want a proof of concept on Windows, but I don't see why it wouldn't
work.

**Comment by RonnyPfannschmidt** [comment]
A hash oft the path element iis required ffor sanity

**Comment by blueyed** [comment]
Ok.

Not really a fan of this myself though - I'd rather keep it with the repository, similar to `.tox` dirs.
Only came here via #4270.

Using a hash would not allow to prune only certain/known dirs for example.
borg appears to use a hash (but you have usually less borg caches than pytest ones).
However, the `CACHEDIR.TAG` file (also used by borg) would be good to have.. (separate issue; haven't searched).

**Comment by RonnyPfannschmidt** [comment]
@blueyed we can make accompanying json files to make accompanying metadata and a prune command

the `CACHEDIR.TAG` should totally be added

in fact we ought to consider adding it even to the .pytest_cache folder like the readme/gitignore

**Comment by nicoddemus** [comment]
This is now possible because `cache_dir` ini option accepts environment variables:

```ini
[pytest]
cache_dir=$XDG_CONFIG_DIR
```

So I believe this can be closed.

**Comment by RonnyPfannschmidt** [comment]
In xdg those have implied defaults, and the example you gave would nuke the entire config dir of the user if it was explicitly set while running cache clean

**Comment by nicoddemus** [comment]
Oh you mean it should have been `cache_dir=$XDG_CONFIG_DIR/pytest-cache`? I see, bad assumption.

But other than that, what else would a user need to use the `$XDG_CONFIG_DIR`?

**Comment by RonnyPfannschmidt** [comment]
Again incorrect, there are implicit defaults in the standard, and the config is NEVER to be used for caches

Additionally there would be confusion about ownership if entirely different projects used the same setup

**Comment by nicoddemus** [comment]
> the config is NEVER to be used for caches

I'm confused... is this issue invalid then?

> Additionally there would be confusion about ownership if entirely different projects used the same setup

Oh definitely I was naive on my quick late night reply.

**Comment by RonnyPfannschmidt** [comment]
The initial comments had the correct var name, the title was wrong, i fixed that

**Comment by nicoddemus** [comment]
> The example above is using a hash key to isolate the cache directories between projects, based on their folder. It is simple but effective.

Ahh right, good idea.

**Comment by merwok** [comment]
I don’t think that CPython uses XDG directories for anything, do you have other info @ssbarnea ?

**Comment by RonnyPfannschmidt** [comment]
@ssbarnea feel free to invest the time to create a proper storage /eviction scheme, which even the named tools only partially implement

**Comment by ssbarnea** [comment]
I see two possible approaches here:
- use only user-level .cache and delay addressing the cleanup for a follow-up contribution as that is not critical (if we include project root folder in hash)
- adopt the project level `.cache` model similar to https://dot-config.github.io/ - basically changing default to be `.cache/pytest_cache` should be enough for that. At least we avoid being another software the clutters our project root directory with temp stuff.

**Comment by feluxe** [comment]
pytest is the only tool in my python stack which still pollutes my source directories. `python`, `poetry`, `pipenv`, etc all use `~/.cache/...` by default or allow me to configure conveniently via env var.

**Comment by SirzBenjie** [comment]
Would also very, very much like to see this, please.

**Comment by 0cjs** [comment]
Using `XDG_CACHE_HOME` would cause confusion and pain. `XDG_*` directories are specifically for users' personal environment configuration (XDG stands for X Desktop Group, and for historical reasons is the term used in the freedesktop.org specifications) and this should _not_ be shared with development projects that may write conflicting information and may want to remove information when doing clean builds and the like.

Consider what can happen if you use `XDG_CONFIG_HOME` for configuration information for a particular project when testing it. Running whatever in your development checkout sets this would overwrite the user's configuration, which may well be different from the configuration used for testing. Further, if you have two different development checkouts, each would overwrite the other's information.

Mixing different projects' build artifacts also makes clean builds more difficult since it's hard to know what to remove and removals may affect other projects. In the current situation, you simply `rm -rf .pytest_cache` or similar in your project dir and you know that a) you are fully clean on that, and b) you have not affected any other projects, or other copies of this project.

I don't mind if you add this as some sort of option for people who really want it, but it should not be the default. 

> pytest is the only tool in my python stack which still pollutes my source directories. python, poetry, pipenv, etc all use ~/.cache/... by default or allow me to configure conveniently via env var.

We have `.gitignore` files for a reason. And env vars are a terrible configuration mechanism because, unlike a `pyproject.toml` or `pytest.ini` file, they are unobvious, easily change from session to session and terminal to terminal, and hard to set reliably in the first place. Where build artifacts go should not be random based on users' configuration, but should be defined in a config file in the project itself so that both users and the build system always know where everything is and always use a consistent set of files.

(FWIW, I just got nailed today by the default 'use $CWD'—effectively the same thing as using an env var—when my `--lf/--ff` options broke because I had two cache directories; which one was used depended on the circumstances. I implemented the correct solution: add a `pytest.ini` file that puts the cache dir in one place for the project, regardless of the my personal environment.)

(Also FWIW, I configure my projects to place _all_ build products, cache dirs, etc. under a `.build/` directory at the top level of the project, at least where the tools don't directly block this technique. This greatly reduces clutter (which I also hate), and also gives you an easy way to do a fully clean build: just remove that one directory.)

**Comment by RonnyPfannschmidt** [comment]
By now there are dozens of tools that map folders in there to worktrees including hatch pre-commit and others 

Having a distinction is a solved problem
